### PR TITLE
Issue 120 

### PR DESCRIPTION
**Pull request by SecTheTech**
- Add support for top alarms : 
```
$ adbe alarm top
android
	Action: TIME_TICK
	Running time: +4s300ms
	Number of device woke up: 0
	Number of alarms: 1973
	User ID: 1000
android
	Action: com.android.server.action.NETWORK_STATS_POLL
	Running time: +3s537ms
	Number of device woke up: 0
	Number of alarms: 66
	User ID: 1000
android
	Action: WifiConnectivityManager Schedule Periodic Scan Timer
	Running time: +770ms
	Number of device woke up: 692
	Number of alarms: 692
	User ID: 1000
android
	Action: WifiConnectivityManager Schedule Watchdog Timer
	Running time: +154ms
	Number of device woke up: 98
	Number of alarms: 98
	User ID: 1000
com.android.providers.calendar
	Action: com.android.providers.calendar.intent.CalendarProvider2
	Running time: +92ms
	Number of device woke up: 6
	Number of alarms: 6
	User ID: u0a42
android
	Action: com.android.server.NetworkTimeUpdateService.action.POLL
	Running time: +71ms
	Number of device woke up: 0
	Number of alarms: 1
	User ID: 1000
com.android.deskclock
	Action: indicator
	Running time: +32ms
	Number of device woke up: 1
	Number of alarms: 1
	User ID: u0a95
android
	Action: ScheduleConditionProvider.EVALUATE
	Running time: +24ms
	Number of device woke up: 3
	Number of alarms: 3
	User ID: 1000
```

### CODE REVIEW

**Review by ashishb — COMMENTED** [review]
This looks good to me broadly. How about removing everything non-alarm related from this PR?

**Comment by SecTheTech** [comment]
I reverted the uninstall function, I will send a pull request separately. 
I think that the current information displayed for Alarms are enough, should we add the rest ? 

- Alarm History
```
$ adbe alarm history
App Alarm history
Package name: com.android.providers.calendar
   User ID:  u0
   history:  -1h35m47s957ms,
Package name: com.android.deskclock
   User ID:  u0
   history:  -2d14h44m13s358ms, -2d14h54m13s357ms, -2d15h24m13s358ms,
```
- Pending Alarms (can be used with the argument `-v`):
```
$ adbe alarm pending 
Pending Alarms:
ID: 1947338
   Number of alarms: 1
   Alarm #0:
      Package: android
ID: be7ffa
   Number of alarms: 1
   Alarm #0:
      Package: android
ID: 1bf8cb4
   Number of alarms: 6
   Alarm #5:
      Package: android
   Alarm #4:
      Package: android
   Alarm #3:
      Package: android
   Alarm #2:
      Package: com.android.providers.calendar
   Alarm #1:
      Package: android
   Alarm #0:
      Package: android
```

```
$ adbe alarm pending -v
             ...
Pending Alarms:
ID: 9b98ae4
   Number of alarms: 1
   Start: 243762373
   End: 243762373
   flag: 0x1
   Alarm #0:
      Type: ELAPSED 
      ID: ca9cb4d
      Type: 3
      When: 243762373
      Package: android
ID: acb3602
   Number of alarms: 2
   Start: 243807807
   End: 243927777
   flag: 0x8
   Alarm #1:
      Type: ELAPSED_WAKEUP 
      ID: 52361c0
      Type: 2
      When: 243807807
      Package: android
   Alarm #0:
      Type: ELAPSED_WAKEUP 
      ID: a86cb13
      Type: 2
      When: 243807777
      Package: android
ID: 354cf49
   Number of alarms: 1
   Start: 244807773
   End: 246157773
   flag: 0x8
   Alarm #0:
      Type: ELAPSED 
      ID: f91449f
      Type: 3
      When: 244807773
      Package: android
ID: dbab04e
   Number of alarms: 1
   Start: 259208042
   End: 259208042
   flag: 0x9
   Alarm #0:
      Type: RTC_WAKEUP 
      ID: 3ed0aeb
      Type: 0
      When: 1609073845670
      Package: android
ID: 489056f
   Number of alarms: 6
   Start: 275196615
   End: 275400000
   flag: 0x8
   Alarm #5:
      Type: ELAPSED 
      ID: 10cd3fa
      Type: 3
      When: 275196615
      Package: android
   Alarm #4:
      Type: ELAPSED 
      ID: 6607406
      Type: 3
      When: 259291376
      Package: android
   Alarm #3:
      Type: ELAPSED_WAKEUP 
      ID: d0f6ac7
      Type: 2
      When: 259211201
      Package: android
   Alarm #2:
      Type: ELAPSED_WAKEUP 
      ID: 558981d
      Type: 2
      When: 259200000
      Package: com.android.providers.calendar
   Alarm #1:
      Type: ELAPSED 
      ID: 87fca92
      Type: 3
      When: 250571246
      Package: android
   Alarm #0:
      Type: ELAPSED_WAKEUP 
      ID: 4860060
      Type: 2
      When: 250571201
      Package: android
```
- All alarms:
Print all the information (top, history and pending)

File: README.md [hunk 1]
```diff
@@ -186,7 +186,7 @@ See [Release announcement](https://ashishb.net/tech/introducing-adb-enhanced-a-s
     adbe [options] dont-keep-activities (on | off)
     adbe [options] animations (on | off)
     adbe [options] show-taps (on | off)
-    adbe [options] stay-awake-while-charging (on | off)
+    adbe [options] stay-awake-while-charging (on | off)
```
Comment (ashishb) [inline README.md:189]: Revert the trailing space.

File: adbe/adb_enhanced.py [hunk 2]
```diff
@@ -1658,21 +1660,20 @@ def alarm_manager(param):
...
                 action = value.split(":")[1]
 
-                print("Package name: %s" % package_name)
-                print("   Action: %s" % action)
-                print("   Running time: %s" % running_time)
-                print("   Number of device woke up: %s" % nb_woke_up)
-                print("   Number of alarms: %s" % nb_alarms)
-                print("   User ID: %s" % uid)
+                print("%sPackage name: %s" % (padding, package_name))
+                print("%s   Action: %s" % (padding * 2, action))
```
Comment (ashishb) [inline adbe/adb_enhanced.py:1664]: Consider using "\t" instead of "   "?

File: adbe/adb_enhanced.py [hunk 3]
```diff
@@ -921,10 +922,14 @@ def _get_all_packages(pm_cmd):
     return packages
 
 
+# https://stackoverflow.com/questions/63416599/adb-shell-pm-list-packages-missing-some-packages
```
Comment (ashishb) [inline adbe/adb_enhanced.py:925]: Please revert from this Pull Request (PR)

File: adbe/adb_enhanced.py [hunk 4]
```diff
@@ -1602,3 +1607,119 @@ def print_notifications():
...
+def alarm_manager(param):
+    cmd = "dumpsys alarm"
+    c, o, e = execute_adb_shell_command2(cmd)
+
+    if c != 0:
+        print_error_and_exit("Something gone wrong on "
+                             "dumping alarms. Error: %s" % e)
+        return o
+
+    if isinstance(param, AlarmEnum):
```
Comment (ashishb) [inline adbe/adb_enhanced.py:1630]: Consider inverting this `if not isinstance(param, AlarmEnum):\n return` to reduce the indentation in the rest of the code

File: adbe/adb_enhanced.py [hunk 5]
```diff
@@ -1602,3 +1607,119 @@ def print_notifications():
...
+
+                print("%sPackage name: %s" % (padding, package_name))
+                print("%s   Action: %s" % (padding * 2, action))
+                print("%s   Running time: %s" % (padding * 2, running_time))
+                print("%s   Number of device woke up: %s" % (padding * 2, nb_woke_up))
+                print("%s   Number of alarms: %s" % (padding * 2, nb_alarms))
+                print("%s   User ID: %s" % (padding * 2, uid))
+
+        if param == AlarmEnum.PENDING or run_all == 1:
+            print("Pending Alarms:")
```
Comment (ashishb) [inline adbe/adb_enhanced.py:1671]: These if branches are long enough that you can extract functions out of them.

**Comment by ashishb** [comment]
> I think that the current information displayed for Alarms are enough, should we add the rest ?

This looks good and complete on its own. You can always add more functionality in the follow-on PRs.

File: adbe/adb_enhanced.py (hunk 2 above)
Comment (SecTheTech) [reply adbe/adb_enhanced.py:1664]: I tried "\t" but it produce a lot of spaces. I will replace it with "\t"

File: adbe/adb_enhanced.py (hunk 5 above)
Comment (SecTheTech) [reply adbe/adb_enhanced.py:1671]: Will update it during the week.

**Comment by SecTheTech** [comment]
I refactored the code, when testing on Android 8 and 9 the logs are not parsed as the format changed in Android 10 (API 29).
This version support only top alarms for API < 29.

File: adbe/adb_enhanced.py [hunk 6]
```diff
@@ -1617,109 +1617,143 @@ class AlarmEnum(Enum):
...
+            print_history_alarms(o, padding)
+        else:
+            print_error("Your Android version does not support " +
+                        "listing pending alarm")
+
+    if param == AlarmEnum.HISTORY or run_all == 1:
+        if api_version > 28:
+            print_history_alarms(o, padding)
+        else:
+            print_error("Your Android version does not support " +
```
Comment (ashishb) [inline adbe/adb_enhanced.py:1752]: Nit: May be add that only API 28 and below are supported in the error message.

**Comment by ashishb** [comment]
> This version support only top alarms for API < 29.

That's OK. It is perfectly fine to not support a limited number of versions if supporting more requires too much effort.

**Comment by ashishb** [comment]
I can approve as soon as https://github.com/ashishb/adb-enhanced/pull/131/files#r549431488 and https://github.com/ashishb/adb-enhanced/pull/131/files#r549432185 are addressed.

**Review by ashishb — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (ashishb) [reply README.md:189]: I will fix this in https://github.com/ashishb/adb-enhanced/pull/133

**Pull request merged**